Adding the first CryptoKit API to webCrypto: AesGcm Encrypt

nmahendru opened the pull request:
#### 2a6d9d74bd17076029b5c2313612086d1009fddb
<pre>
Adding the first CryptoKit API to webCrypto: AesGcm Encrypt
<a href="https://bugs.webkit.org/show_bug.cgi?id=246350">https://bugs.webkit.org/show_bug.cgi?id=246350</a>
<a href="https://rdar.apple.com/101040216">rdar://101040216</a>

Reviewed by Alex Christensen and Elliott Williams.

This PR does the following:
1. Adds plumbing to call into CryptoKit which is swift only.
2. Only adds one API for now just to gather early feedback on how it&apos;s structured.
3. Tries to compare the encryption outputs with the existing implementation bit for bit
   to prove that this works exactly the same.

* LayoutTests/crypto/subtle/aes-gcm-encrypt-malformed-parameters-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-encrypt-malformed-parameters.html:
* LayoutTests/crypto/subtle/aes-gcm-generate-key-encrypt-decrypt-null-plain-text-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-generate-key-encrypt-decrypt-null-plain-text.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-tag-length-32-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-tag-length-32.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-tagLengths-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-tagLengths.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Configurations/WebCoreSwift.xcconfig: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/WebCoreSwift/CryptoKitShim.swift: Added.
* Source/WebCore/crypto/CryptoAlgorithm.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAES_GCM.cpp:
(WebCore::CryptoAlgorithmAES_GCM::encrypt):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAES_GCM.h:
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmAES_GCMGCrypt.cpp:
(WebCore::CryptoAlgorithmAES_GCM::platformEncrypt):
* Source/WebCore/crypto/mac/CryptoAlgorithmAES_GCMMac.mm:
(WebCore::encryptAES_GCM):
(WebCore::encryptCryptoKitAES_GCM):
(WebCore::CryptoAlgorithmAES_GCM::platformEncrypt):
* Source/WebCore/crypto/openssl/CryptoAlgorithmAES_GCMOpenSSL.cpp:
(WebCore::CryptoAlgorithmAES_GCM::platformEncrypt):
* Source/WebCore/WebCorePrefix.h:

Canonical link: <a href="https://commits.webkit.org/273488@main">https://commits.webkit.org/273488@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c5684c9df4041996b864351c11ca1514d33e6dd2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35503 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14449 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37640 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38245 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32008 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36701 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16830 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11482 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30836 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36055 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12205 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31614 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10720 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10746 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31739 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39492 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/30104 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32267 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32071 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36713 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/35387 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10918 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8824 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34766 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12639 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/31402 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/42060 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8123 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11435 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8731 "Found 39 new JSC stress test failures: microbenchmarks/array-from-derived-object-func.js.bytecode-cache, microbenchmarks/array-from-derived-object-func.js.default, microbenchmarks/array-from-derived-object-func.js.dfg-eager, microbenchmarks/array-from-derived-object-func.js.dfg-eager-no-cjit-validate, microbenchmarks/array-from-derived-object-func.js.eager-jettison-no-cjit, microbenchmarks/array-from-derived-object-func.js.no-cjit-collect-continuously, microbenchmarks/array-from-derived-object-func.js.no-cjit-validate-phases, microbenchmarks/array-from-derived-object-func.js.no-llint, microbenchmarks/array-from-object-func.js.bytecode-cache, microbenchmarks/array-from-object-func.js.default ... (failure)") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11702 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -44132,13 +44230,20 @@
+				CLANG_ENABLE_MODULES = YES;
+				DEFINES_MODULE = YES;
+				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
+				SWIFT_VERSION = 5.0;
```
Doesn't seem like this patch adds Swift code to WebCore, so why do we need these settings?

nmahendru replied:
as discussed in person, this is adding `CryptoKitShim.swift` to WebCore.

emw-apple replied:
My understanding is that these settings are what let WebCore import the `WebCoreSwift` swiftmodule (not that a Swift source is being added to WebCore).

Also, having WebCore setting `CLANG_ENABLE_MODULES` likely has build performance impact. I'm somewhat surprised it works at all without additional tweaking. Are you sure we need it?

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -53,7 +53,10 @@ FRAMEWORK_SEARCH_PATHS[sdk=macosx*] = $(WK_QUOTED_OVERRIDE_FRAMEWORKS_DIR);
 
-HEADER_SEARCH_PATHS = PAL ForwardingHeaders /usr/include/libxslt /usr/include/libxml2 "$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore" "$(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH)" $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(ANGLE_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) $(HEADER_SEARCH_PATHS) $(SRCROOT);
+// FIXME: Is there a better way with lesser hardcoded values?
+WEBCORE_SWIFT_HEADERS = "$(DERIVED_DATA_DIR)"/WebCore.build/$(CONFIGURATION)/WebCoreSwift.build/DerivedSources
```
I think you're looking for $(DERIVED_SOURCES_DIR). Although I would have expected that it's on the headers search path by default.

nmahendru replied:
`CONFIGURATION_TEMP_DIR` is what worked in the end for all platforms and builds we have.
What you are referring to(if I am not wrong) is the directory used for same target includes.
This is cross target.

emw-apple replied:
Yeah, I was misreading. Thank you!

emw-apple reviewed (commented):
Some early feedback: You'll want to move all build settings out of the pbxproj and into xcconfig files. That will probably entail adding a new `WebCoreSwift.xcconfig`.

achristensen07 commented on Source/WebCore/crypto/mac/cryptokit/CryptoKitAesShim.mm:
```diff
@@ -0,0 +1,47 @@
+namespace WebCore {
+
+int CkAesGcm::encrypt(
+    const uint8_t * key,
```
WebKit style has no space before *, usually all parameters on same line, same in header.

nmahendru replied:
fixed.

achristensen07 commented on Source/WebCore/crypto/mac/CryptoAlgorithmAES_GCMMac.cpp:
```diff
@@ -69,8 +78,11 @@ ALLOW_DEPRECATED_DECLARATIONS_END
 }
 
-ExceptionOr<Vector<uint8_t>> CryptoAlgorithmAES_GCM::platformEncrypt(const CryptoAlgorithmAesGcmParams& parameters, const CryptoKeyAES& key, const Vector<uint8_t>& plainText)
+ExceptionOr<Vector<uint8_t>> CryptoAlgorithmAES_GCM::platformEncrypt(const CryptoAlgorithmAesGcmParams& parameters, const CryptoKeyAES& key, const Vector<uint8_t>& plainText, bool useCryptoKit)
```
Instead of adding a bool parameter, we may want to add a new CryptoAlgorithmAES_GCM that only uses CryptoKit.  It's nbd to add a bool in one place, but as we expand this it may be structured better to just have a new implementation object.

nmahendru replied:
As discussed:
1. bool will still be required to select between the two at runtime.
2. Once I make CK the default this bool will just go away.

achristensen07 commented on Source/WebCore/WebCoreSwift/CryptoKitShim.swift:
```diff
@@ -0,0 +1,94 @@
+                return AESGCMErrorCodes.TooBigArguments.rawValue
+            }
+            if ivSize == 0 {
+                    return AESGCMErrorCodes.InvalidArgument.rawValue
```
Looks like too much indentation.

nmahendru replied:
fixed.

achristensen07 commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -2216,6 +2216,20 @@ DeprecateAESCFBWebCryptoEnabled:
     WebCore:
       default: true
 
+DeprecateCoreCommonCryptoEnabled:
```
I think "CryptoKitEnabled" would be a better name for this.  It avoids double negatives when reading code.

nmahendru replied:
done

achristensen07 commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -53,7 +53,10 @@ FRAMEWORK_SEARCH_PATHS[sdk=macosx*] = $(WK_QUOTED_OVERRIDE_FRAMEWORKS_DIR);
 SYSTEM_FRAMEWORK_SEARCH_PATHS = $(inherited) $(WK_PRIVATE_SDK_DIR)$(SYSTEM_LIBRARY_DIR)/PrivateFrameworks $(SDK_DIR)$(SYSTEM_LIBRARY_DIR)/Frameworks;
 
-HEADER_SEARCH_PATHS = PAL ForwardingHeaders /usr/include/libxslt /usr/include/libxml2 "$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore" "$(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH)" $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(ANGLE_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) $(HEADER_SEARCH_PATHS) $(SRCROOT);
+// FIXME: Is there a better way with lesser hardcoded values?
```
This comment seems unnecessary.  The answer is probably pretty close to no.

nmahendru replied:
You are right. Removed now.

achristensen07 commented on Source/WebCore/Configurations/WebCoreSwift.xcconfig:
```diff
@@ -0,0 +1,33 @@
+// Copyright (C) 2009-2024 Apple Inc. All rights reserved.
```
Why 2009?  This is a new file

nmahendru replied:
ah.. copy and paste from another config file. Fixed.

achristensen07 commented on Source/WebCore/crypto/mac/CryptoAlgorithmAES_GCMMac.cpp:
```diff
@@ -43,12 +44,20 @@ ALLOW_DEPRECATED_DECLARATIONS_BEGIN
 
+static ExceptionOr<Vector<uint8_t>> encryptCryptoKitAES_GCM(const Vector<uint8_t>& iv, const Vector<uint8_t>& key, const Vector<uint8_t>& plainText, const Vector<uint8_t>& additionalData, size_t desiredTagLengthInBytes)
+{
+    Vector<uint8_t> cipherText(plainText.size() + 16);
```
Let's give 16 a name so we don't have any magic numbers.  It looks like some kind of header size.

nmahendru replied:
done.

achristensen07 commented on Source/WebCore/crypto/mac/CryptoAlgorithmAES_GCMMac.cpp:
```diff
@@ -43,12 +44,20 @@ ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    Vector<uint8_t> cipherText(plainText.size() + 16);
+    if (CkAesGcm::encrypt(key.data(), key.size(), iv.data(), iv.size(), additionalData.data(), additionalData.size(), plainText.data(), plainText.size(), cipherText.data()))
+        return Exception { ExceptionCode::OperationError };
+    cipherText.resize(cipherText.size() - (16 - desiredTagLengthInBytes));
```
What if desiredTagLengthInBytes is more than 16?

nmahendru replied:
This actually is checked one layer above. But I have added another check here regardless. Have also unified this 16 value in a constant and used that everywhere.

achristensen07 commented on Source/WebCore/crypto/mac/cryptokit/CryptoKitAesShim.h:
```diff
@@ -0,0 +1,47 @@
+
+class CkAesGcm {
+public:
+    static int encrypt(
```
Can we return an error enum instead of just an int?
Do we want to handle any errors?
Can we pass std::span<const uint8_t> here instead of pointer/size?
It seems concerning that we pass a ciphertext pointer but not a size of that.

nmahendru replied:
> It seems concerning that we pass a ciphertext pointer but not a size of that.

for AES_GCM plaintext length always is equal to ciphertext length. Just reducing one parameter.

> Do we want to handle any errors?

It's either a pass or a fail for this one. There is nothing to be handled and taken action on. The implementation before does the same.

> Can we pass std::span here instead of pointer/size?

I have removed the shim completely and also now passing in a reference to the original vectors.

emw-apple commented on Source/WebCore/crypto/CryptoAlgorithm.h:
```diff
@@ -31,6 +31,7 @@
 #include "CryptoKeyUsage.h"
 #include "ExceptionOr.h"
 #include "JsonWebKey.h"
+#include "ScriptExecutionContext.h"
```
Curious why this is needed, since nothing in the header has changed to require more type information about `ScriptExecutionContext`. Is it to fix the unified build?

nmahendru replied:
I have now added a forward decl for this. And added the header in another file where I needed it.

emw-apple commented on Source/WebCore/crypto/mac/CryptoAlgorithmAES_GCMMac.cpp:
```diff
@@ -43,12 +44,20 @@ ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+static ExceptionOr<Vector<uint8_t>> encryptCryptoKitAES_GCM(const Vector<uint8_t>& iv, const Vector<uint8_t>& key, const Vector<uint8_t>& plainText, const Vector<uint8_t>& additionalData, size_t desiredTagLengthInBytes)
+{
+    Vector<uint8_t> cipherText(plainText.size() + 16);
+    if (CkAesGcm::encrypt(key.data(), key.size(), iv.data(), iv.size(), additionalData.data(), additionalData.size(), plainText.data(), plainText.size(), cipherText.data()))
```
Since this file is already Cocoa-only, can we turn it into an ObjC++ file (.mm) and call the Swift shim directly? That ought to eliminate the C++ shim in `CryptoKitAesShim.{h,mm}`, so we'll only need one shim instead of two :)

nmahendru replied:
great idea. Relentless code reduction appetite, I like it!

nmahendru replied:
done. the `mm/h` shim is gone now.

emw-apple commented on Source/WebCore/WebCoreSwift/CryptoKitShim.swift:
```diff
@@ -0,0 +1,94 @@
+        _ additionalDataSize: UInt,
+        _ plaintext: UnsafePointer<UInt8>,
+        _ plaintextSize: UInt,
+        _ ciphertext: UnsafeMutablePointer<UInt8>) -> Int
```
I would prefer having argument labels here. As is, the generated header is weird (lots of empty parameter names like `+[WKAesGcm encrypt:x :y :z ...]`), and since the types are all so primitive, it seems useful to have the lablels.

nmahendru replied:
done.

emw-apple commented on Source/WebCore/WebCoreSwift/CryptoKitShim.swift:
```diff
@@ -0,0 +1,94 @@
+        } catch {
+            return AESGCMErrorCodes.EncryptionFailed.rawValue
+        }
+        return AESGCMErrorCodes.Success.rawValue
```
Echoing Alex, I think it should be fine to return enum cases here, since the `@objc` annotation on the enum itself will translate into an NS_ENUM.

nmahendru replied:
done.

emw-apple commented on Source/WebCore/WebCorePrefix.h:
```diff
@@ -35,9 +35,9 @@
 #ifdef __cplusplus
-#define NULL __null
+#include <cstddef>
 #else
-#define NULL ((void *)0)
+#include <stddef.h>
```
What do we need these for? It's a lot to pull into the prefix header.

nmahendru replied:
the error that I am trying to fix here is that this definition is confliciting with stddef.h so to permanently fix this I am just letting the header deal with it. It's anyways included further down in another file with which it is conflicting with.

emw-apple replied:
Interesting… why is this an issue now? Is it because we've turned on clang modules? (This change seems fine, but just curious)

nmahendru replied:
AFAIK...Unified Sources do expose issues like this whilst adding new files/reordering files.

emw-apple commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -42947,6 +43008,23 @@
+			name = WebCoreSwift;
+			productName = WebCoreSwift;
+			productReference = 948FD0392B4C9D7A00043FEB /* libWebCoreSwift.dylib */;
+			productType = "com.apple.product-type.library.dynamic";
```
Can we make this a static library? As a dylib, we would need to ensure it gets installed properly in the OS build, or WebKit won't be able to launch (i.e. `SKIP_INSTALL` is certainly wrong). It may also have a larger binary size impact.

As a static library, the code gets merged into the final WebCore binary and the library (and associated swiftmodule) can be discarded, so it's much easier to deal with from a build architecture standpoint.

It's really annoying to change library type in Xcode but I think you can just change the produce type to `com.apple.product-type.library.static` and replace all `.dylib` references with `.a`.

nmahendru replied:
changed it to static lib.

emw-apple commented on Source/WebCore/Configurations/WebCoreSwift.xcconfig:
```diff
@@ -0,0 +1,31 @@
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
+
+PRODUCT_NAME = WebCoreSwift;
+SWIFT_OBJC_INTERFACE_HEADER_NAME = WebCoreSwift-Swift.h;
```
We need to add `SKIP_INSTALL = YES`. Since this library gets merged into WebCore, it should not be installed to the system.

nmahendru replied:
done

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -174,7 +176,7 @@ JSC_SEC_LD_FLAGS[sdk=watch*] = ;
 
 // FIXME: Reduce the number of allowable_clients <rdar://problem/31823969>
-OTHER_LDFLAGS = $(inherited) $(WK_RELOCATABLE_FRAMEWORK_LDFLAGS) $(JSC_SEC_LD_FLAGS) -weak-lxslt -lsqlite3 -lobjc -allowable_client WebCoreTestSupport -allowable_client WebKitLegacy -allowable_client WebKit -allowable_client TestIPC -allowable_client TestWebKitAPI -allowable_client DumpRenderTree -allowable_client WebKitTestRunner -force_load $(BUILT_PRODUCTS_DIR)/libPAL.a -framework CFNetwork -f ...
+OTHER_LDFLAGS = $(inherited) $(WK_RELOCATABLE_FRAMEWORK_LDFLAGS) $(JSC_SEC_LD_FLAGS) -weak-lxslt -lsqlite3 -lobjc -allowable_client WebCoreTestSupport -allowable_client WebKitLegacy -allowable_client WebKit -allowable_client TestIPC -allowable_client TestWebKitAPI -allowable_client DumpRenderTree -allowable_client WebKitTestRunner -force_load $(BUILT_PRODUCTS_DIR)/libPAL.a $(BUILT_PRODUCTS_DIR)/l ...
```
For consistency and extra caution, use -force_load here like we do with PAL:
```suggestion
OTHER_LDFLAGS = $(inherited) $(WK_RELOCATABLE_FRAMEWORK_LDFLAGS) $(JSC_SEC_LD_FLAGS) -weak-lxslt -lsqlite3 -lobjc -allowable_client WebCoreTestSupport -allowable_client WebKitLegacy -allowable_client WebKit -allowable_client TestIPC -allowable_client TestWebKitAPI -allowable_client DumpRenderTree -allowable_client WebKitTestRunner -force_load $(BUILT_PRODUCTS_DIR)/libPAL.a -force_load $(BUILT_PRODUCTS_DIR)/libWebCoreSwift.a -lswiftCore -framework CFNetwork -framework CoreAudio -framework CoreGraphics -framework CoreText -framework Foundation -framework IOSurface -framework ImageIO -framework Metal -framework Network -lFontParser $(OTHER_LDFLAGS_PLATFORM_$(WK_COCOA_TOUCH)) $(OTHER_LDFLAGS_PLATFORM_$(WK_PLATFORM_NAME)) $(WK_ANGLE_LDFLAGS) $(WK_WEBGPU_LDFLAGS) $(WK_APPKIT_LDFLAGS) $(WK_APPSUPPORT_LDFLAGS) $(WK_AUDIO_UNIT_LDFLAGS) $(WK_CARBON_LDFLAGS) $(WK_CORE_UI_LDFLAGS) $(WK_DATA_DETECTORS_CORE_LDFLAGS) $(WK_GRAPHICS_SERVICES_LDFLAGS) $(WK_IOSURFACE_ACCELERATOR_LDFLAGS) $(WK_LIBWEBRTC_LDFLAGS) $(WK_MOBILE_CORE_SERVICES_LDFLAGS) $(WK_MOBILE_GESTALT_LDFLAGS) $(WK_NETWORK_EXTENSION_LDFLAGS) $(WK_SYSTEM_CONFIGURATION_LDFLAGS) $(WK_CORE_IMAGE_LDFLAGS) $(WK_URL_FORMATTING_LDFLAGS) $(WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS) $(WK_XR_RUNTIME_SUPPORT_LDFLAGS) $(WK_SCENEKIT_LDFLAGS) $(SOURCE_VERSION_LDFLAGS) $(PROFILE_GENERATE_OR_USE_LDFLAGS) $(WK_NO_STATIC_INITIALIZERS) $(WK_APPLEJPEGXL_LDFLAGS) $(WK_DEBUG_LDFLAGS);
```

That will prevent the linker from trying to dead-strip code from WebCoreSwift, which could happen in the future if e.g. a swift file only containing an extension were added.

emw-apple commented on Source/WebCore/Configurations/WebCoreSwift.xcconfig:
```diff
@@ -0,0 +1,33 @@
+
+SWIFT_INSTALL_OBJC_HEADER = YES;
+SWIFT_VERSION = 5.0;
+SWIFT_LINT = YES;
```
I don't think `SWIFT_LINT` is a real thing in our codebase.

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -206,3 +208,5 @@ WK_HAVE_URL_FORMATTING_appletvsimulator = NO;
 WK_HAVE_URL_FORMATTING_macosx = $(WK_HAVE_URL_FORMATTING$(WK_MACOS_1014));
 WK_HAVE_URL_FORMATTING_MACOS_SINCE_1014 = YES;
+
+SWIFT_LINT = YES;
```
Ditto.

emw-apple commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -21297,13 +21324,16 @@
 				1C8D26D022C09CDE00D125F3 /* libcompression.tbd */,
 				93F1D31A0558CC5C00821BC0 /* libicucore.dylib */,
 				1CFAE3220A6D6A3F0032593D /* libobjc.dylib */,
+				94EDD04C2B59E8B3001A3AC4 /* libswiftCore.tbd */,
```
I don't think we need a dependency on libswift added in Xcode's UI if we also specify the dependency in OTHER_LDFLAGS.

emw-apple commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -21297,13 +21324,16 @@
 				A1763F401E20586E001D58DE /* CFNetwork.framework */,
 				AA5F3B9016CC5BEB00455EB0 /* CoreFoundation.framework */,
+				948FD0332B4C9D0600043FEB /* CryptoKit.framework */,
+				948FD0432B4C9D8800043FEB /* CryptoKit.framework */,
```
Looks like an extra link of CryptoKit, please remove.

emw-apple commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -29493,6 +29524,21 @@
+			);
+			path = cryptokit;
+			sourceTree = "<group>";
+		};
```
Looks like a reference to a nonexistent directory.

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -53,9 +53,11 @@ FRAMEWORK_SEARCH_PATHS[sdk=macosx*] = $(WK_QUOTED_OVERRIDE_FRAMEWORKS_DIR);
-HEADER_SEARCH_PATHS = PAL ForwardingHeaders /usr/include/libxslt /usr/include/libxml2 "$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore" "$(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH)" $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(ANGLE_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) $(HEADER_SEARCH_PATHS) $(SRCROOT);
+WEBCORE_SWIFT_HEADERS = $(CONFIGURATION_TEMP_DIR)/WebCoreSwift.build/DerivedSources
+
+HEADER_SEARCH_PATHS = PAL ForwardingHeaders /usr/include/libxslt /usr/include/libxml2 "$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore" "$(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH)" $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(ANGLE_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) "$(WEBCORE_SWIFT_HEADERS)" $(HEADER_SEARCH_PATHS) $(SRCROOT);
```
The double-quotes around some variables are misleading; this does nothing in XCBuild, and build settings do correctly expand to include spaces. Since you're touching this line, would you mind removing all the quotes?

emw-apple commented on Source/WebCore/Configurations/WebCoreSwift.xcconfig:
```diff
@@ -0,0 +1,33 @@
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
+
+PRODUCT_NAME = WebCoreSwift;
+SWIFT_OBJC_INTERFACE_HEADER_NAME = WebCoreSwift-Swift.h;
```
I wonder if it would be more natural to call this `WebCore-Swift.h`

nmahendru replied:
I am just following the convention for the generatedHeaders.

emw-apple reviewed (commented):
Here's some more feedback. I'm happy with the overall approach, so I'm being a bit pickier with the build changes to get this ready to land.

emw-apple commented on Source/WebCore/WebCoreSwift/CryptoKitShim.swift:
```diff
@@ -0,0 +1,94 @@
+            if ivSize == 0 {
+                return AesGcmErrorCodes.InvalidArgument
+            }
+            let zeroArray = [UInt8](repeating: 0, count: 0);
```
nit: unnecessary semicolon
```suggestion
            let zeroArray = [UInt8](repeating: 0, count: 0)
```

emw-apple replied:
Also, out of curiosity, why do we need this at all? If `plainTextSize` is `0`, it doesn't seem like it would matter which start address we pass in (the address of `plainText` or `zeroArray`), because the size will be 0 and it won't be dereferenced.

nmahendru committed:
Adding the first CryptoKit API to webCrypto: AesGcm Encrypt
https://bugs.webkit.org/show_bug.cgi?id=246350
rdar://101040216

Reviewed by Alex Christensen and Elliott Williams.

This PR does the following:
1. Adds plumbing to call into CryptoKit which is swift only.
2. Only adds one API for now just to gather early feedback on how it's structured.
3. Tries to compare the encryption outputs with the existing implementation bit for bit
   to prove that this works exactly the same.

* LayoutTests/crypto/subtle/aes-gcm-encrypt-malformed-parameters-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-encrypt-malformed-parameters.html:
* LayoutTests/crypto/subtle/aes-gcm-generate-key-encrypt-decrypt-null-plain-text-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-generate-key-encrypt-decrypt-null-plain-text.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-tag-length-32-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-tag-length-32.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-tagLengths-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-tagLengths.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Configurations/WebCoreSwift.xcconfig: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/WebCoreSwift/CryptoKitShim.swift: Added.
* Source/WebCore/crypto/CryptoAlgorithm.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAES_GCM.cpp:
(WebCore::CryptoAlgorithmAES_GCM::encrypt):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAES_GCM.h:
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmAES_GCMGCrypt.cpp:
(WebCore::CryptoAlgorithmAES_GCM::platformEncrypt):
* Source/WebCore/crypto/mac/CryptoAlgorithmAES_GCMMac.mm:
(WebCore::encryptAES_GCM):
(WebCore::encryptCryptoKitAES_GCM):
(WebCore::CryptoAlgorithmAES_GCM::platformEncrypt):
* Source/WebCore/crypto/openssl/CryptoAlgorithmAES_GCMOpenSSL.cpp:
(WebCore::CryptoAlgorithmAES_GCM::platformEncrypt):
* Source/WebCore/WebCorePrefix.h:

Canonical link: https://commits.webkit.org/273488@main
